packages/amazon-sumerian-hosts-core/src/core/awspack/LexFeature.js

```diff
@@ -0,0 +1,287 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+import Deferred from 'core/Deferred';
+import Utils from 'core/Utils';
+import LexUtils from 'core/awspack/LexUtils';
+
+/**
+ * The AWS LexRuntime service object.
+ * @external LexRuntime
+ * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/LexRuntime.html
+ */
+
+/**
+ * Feature class for interacting with Lex bot which the response from lex bot can be used for speech or other purpose.
+ *
+ * @property {Object} LEX_DEFAULTS - Default values to use with calls to {@link external:LexRuntime}.
+ * @property {string} [LEX_DEFAULTS.SampleRate='16000']
+ * @property {Object} EVENTS - Built-in messages that the feature emits.
+ * @property {string} [EVENTS.lexResponse=onLexResponseEvent] - Message that is emitted after
+ * receiving lex response for the input sent
+ * @property {string} [EVENTS.micReady=onMicrophoneReadyEvent] - Message that is emitted after
+ * microphone is ready to use
+ * @property {string} [EVENTS.micReady=onMicrophoneBeginRecordingEvent] - Message that is emitted after
+ * microphone starts recording
+ * @property {string} [EVENTS.micReady=onMicrophoneEndRecordingEvent] - Message that is emitted after
+ * microphone ends recording
+ * @property {Object} SERVICES - AWS services that are necessary for the feature
+ * to function.
+ * @property {external:LexRuntime} SERVICES.lexRuntime - The LexRuntime service that is used
+ * to interact with lex bot. Will be undefined until [initializeService]{@link LexFeature.initializeService}
+ * has been successfully executed
+ */
+class LexFeature{
+  /**
+   * @constructor
+   *
+   * @param {Messenger=} messenger - Optional Messenger object which the event will be emitted through
+   * @param {Object=} options - Options that will be used to interact with lex bot.
+   * @param {string=} options.botName - The name of the lex bot.
+   * @param {string=} options.botAlias - The alias of the lex bot.
+   * @param {string=} options.userId - The userId used to keep track of the session with lex bot. 
+  */
+  constructor(
+    messenger = undefined,
+    options = {
+      botName: undefined,
+      botAlias: undefined,
+      userId: undefined,
+    }
+  ) {
+    this._messenger = messenger;
+
+    this._botName = options.botName;
+    this._botAlias = options.botAlias;
+    this._userId = options.userId ? options.userId : Utils.createId();
+
+    //Microphone related fields
+    this._recording = false;
+    this._recLength = 0;
+    this._recBuffer = [];
+    this._setAudioContext();
+  }
+
+  /**
+   * Store LexRuntime and AWS SDK Version for use across all instances.
+   *
+   * @param {external:LexRuntime} lexRuntime - lexRuntime instance to use to interact with lex bot.
+   */
+  static initializeService(lexRuntime) {
```

Comment:
Is there a particular reason this is static? In my experience, any kind of static state can lead to big problems when any implicit assumptions are made.

For example, in this code the lex runtime is `this.SERVICES.lexRuntime`. Does this mean that a customer can only use a single instance of the lexruntime at one? I agree that it's not obvious why this might not work for customers, but it also seems arbitrarily constrained. There are mechanically valid, if unusual, usecases like using lex's bots from multiple accounts or regions at the same that are not possible with this setup right?

Reply:
I could make this a member variable of LexFeature instance. However, since i think the config of LexRuntime comes from the AWS object on the page, i'm not sure if multiple accounts/regions are allowed at the same time?